_posts/2018-11-07-retired-pitfalls-22.md

```diff
@@ -0,0 +1,153 @@
+another function or even return it.
+
+This is cool but a bit of a headache for the compiler. To create a fun we pass
+its definition and free variables to a `make_fun2` instruction, but we can't
+include the fun itself as a free variable because it hasn't been created yet.
+Prior to OTP 22 we solved this by creating a new equivalent fun _inside_ the
+fun, but this made recursion surprisingly expensive both in terms of run-time
+and memory use.
+
+As of OTP 22 we translate recursion to a direct function call instead which
+avoids creating a new fun. Other cases still require recreating the fun, but
+they're far less common.
+
+[Optimize named funs and fun-wrapped macros #1973](https://github.com/erlang/otp/pull/1973)
+
+### List subtraction with large operands (-- operator)
+
+While the Erlang VM appears to be pre-emptively scheduled to the programmer,
+it's co-operatively scheduled internally. When a native function runs it
+monopolizes the scheduler until it returns so a long-running one can severely
+harm the responsiveness of the system. We've therefore written nearly all such
+functions in a style that breaks the work into short units that complete
+quickly enough, but there's a steadily shrinking list of functions that
+misbehave, and list subtraction was one of these.
+
+It's usually pretty straightforward to rewrite functions in this style, but
+the old algorithm processed the second list in a loop around the first list,
+which is problematic since both lists can be very long and resuming work in
+nested loops is almost always trickier than expected.
+
+In this case it was easier to just get rid of the nested loop altogether. The
+new algorithm starts out by building a red-black tree from the right-hand side
+before removing elements from the left-hand side. As all operations on the tree
+have `log n` complexity we know that they will finish really quickly, so all we
+need to care about is yielding in the outer loops.
+
+This also had the nice side-effect of reducing the run-time complexity from
+`n * m` to `n log n` and let us remove some warnings from the reference manual
+and [efficiency guide](http://erlang.org/documentation/doc-10.1/doc/efficiency_guide/commoncaveats.html#operator-----);
+the new implementation is always faster than the old workarounds.
+
+This change will be rolled out in OTP 21.2, big thanks to
+Dmytro Lytovchenko (@kvakvs on GitHub) for writing the better half of it!
+
+[Optimize list subtraction (A -- B) and make it yield on large inputs #1998](https://github.com/erlang/otp/pull/1998)
+
+### Lookahead in bit-syntax matching
+
+The optimization pass for bit-syntax matching was completely rewritten in OTP
+22 to take advantage of the new SSA-based intermediate format. It applies the
+same optimizations as before so already well-optimized code is unlikely to see
+any benefit, but it manages to apply them in far more cases.
+
+For those who aren't familiar, all bit-syntax matching operates on a "match
```

Comment:
Perhaps add a link to the Efficiency Guide for more information about match contexts.